File: Assets/Scripts/Spells/FireBolt.js
#pragma strict
class FireBolt extends Spell {
	var cost : float = 1; // cost is required for all spells

	var speed : float = 2;
	var prefab : GameObject = Resources.Load("FireBolt", GameObject);
	
	private var instance : GameObject;
	
	function Cast(playerTrans : Transform) {
		Debug.Log('firing firebolt..');
	
		// test for collision
		var ray = Camera.main.ScreenPointToRay(Input.mousePosition);
		var hit : RaycastHit;
		var targetPos : Vector3 = (playerTrans.position + (playerTrans.forward * 100));
		if (Physics.Raycast(ray, hit, 100))
		{
			targetPos = hit.point;
			if (hit.collider.gameObject.tag == "mob")
			{
				var info = new SpellInfo();
				info.damage = GetDamage();
				info.element = "Fire";
				hit.collider.gameObject.SendMessage("OnSpellCast", info);
			}
		}
	
		// create the particles/objects
		instance = GameObject.Instantiate(prefab, playerTrans.position, playerTrans.rotation) as GameObject;
		instance.audio.Play();
		var instStraightScript : Straight = instance.GetComponent(Straight);
		instStraightScript.speed = speed;
		instStraightScript.target = targetPos;
		instStraightScript.moving = true;
		
	}
	
	function GetDamage(){
		return 1;
	}
}

File: Assets/Scripts/Spells/FireBolt.cs
using UnityEngine;
using System.Collections;

public class FireBolt : Spell
{
    [JsonFx.Json.JsonIgnoreAttribute]
    public int Cost = 1;
    [JsonFx.Json.JsonIgnoreAttribute]
    public float Speed = 2F;
    [JsonFx.Json.JsonIgnoreAttribute]
    public GameObject Prefab = Resources.Load("FireBolt", typeof (GameObject)) as GameObject;
    private GameObject _instance;

    public override string Name {
        get {
            return "FireBolt";
        }
        set {
            base.Name = value;
        }
    }

    public override bool OnFire(Transform playerTransform, Character character)
    {
        Ray ray = Camera.main.ScreenPointToRay(Input.mousePosition); // TODO: always mouse pos?
        RaycastHit hit;
        Vector3 targetPos = playerTransform.position + (playerTransform.forward * 100);

        // test for collision
        if (Physics.Raycast(ray, out hit, 100F))
        {
            targetPos = hit.point;
            if (hit.collider.gameObject.tag == "mob")
            {
                var info = new SpellInfo { Damage = GetDamage(), Element = "Fire" };
                hit.collider.gameObject.SendMessage("OnSpellCast", info);
            }
        }

        // create particles
        _instance = Object.Instantiate(Prefab, playerTransform.position, playerTransform.rotation) as GameObject;
        if (_instance != null)
        {
            _instance.audio.Play();
            var instanceStraightScript = _instance.GetComponent<Straight>();
            instanceStraightScript.Speed = Speed;
            instanceStraightScript.Target = targetPos;
            instanceStraightScript.Moving = true;
        }


        return true;
    }

    private float GetDamage()
    {
        return 1.0F;
    }


}
